core/event/manager.go

```diff
@@ -0,0 +1,30 @@
+// Package event provides a basic API for app modules to emit events.
+package event
+
+import (
+	"context"
+
+	"google.golang.org/protobuf/runtime/protoiface"
+)
+
+// Service represents an event service which can retrieve and set an event manager in a context.
+// event.Service is a core API type that should be provided by the runtime module being used to
+// build an app via depinject.
+type Service interface {
+	// GetEventManager returns the event manager for the context or a no-op event manager if one isn't attached.
+	GetEventManager(context.Context) Manager
+}
+
+// Manager represents an event manager.
+type Manager interface {
+	// EmitProtoEvent emits events represented as a protobuf message (as described in ADR 032).
+	EmitProtoEvent(event protoiface.MessageV1) error
+
+	// EmitKVEvent emits an event based on an event and kv-pair attributes.
+	EmitKVEvent(eventType string, attrs ...KVEventAttribute) error
+}
```

Comment:
Should we maybe do something like this instead?
```suggestion
type Service interface {
	// EmitProtoEvent emits events represented as a protobuf message (as described in ADR 032).
	EmitProtoEvent(ctx context.Context, event protoiface.MessageV1) error

	// EmitKVEvent emits an event based on an event and kv-pair attributes.
	EmitKVEvent(ctx context.Context, eventType string, attrs ...KVEventAttribute) error
}
```

The main reason not to do this I think would be if we want to allow overriding the event manager, i.e. with a method like `WithEventManager(context.Context, EventManager) context.Context` on `event.Service`. I'm not sure if we want to support that or not.

Reply:
I am pro having ctx for consistency.

Reply:
So rather than having methods in `KVStore` to branch in core (which currently don't exist), we have some sort of "cache service" which has a method like `CacheContext(context.Context) (context.Context, CacheWriter)` which branches all stores, events, etc.? Is that what you're thinking @amaurym ?

Reply:
yes, I am in line with @amaurym thinking here, I think branching makes more sense context wise

Reply:
Any thoughts here @alexanderbez or @tac0turtle ?

Reply:
I don have a strong preference here. Im fine with either

Reply:
Whatever we decide, I think it should be in line with https://github.com/cosmos/cosmos-sdk/pull/14444, which isn't merged but I think is a very solid approach.

Reply:
I think the overall goal would that apps compose these service, we provide a default but if someone wants to override it, they would compose a runtime or application and use their own event service that uses this interface. This approach of composability is cleaner in my mind, not sure if others agree?

Reply:
> One fundamental question is do we want to allow users to override the event manager in the context for any purposes other than caching? Is there a use case for that?

No, I can't think of a good one. I do recall there being one PR around the event manager where someone overrode it I think.

Reply:
I would say service is fine here. that is all that's needed to override it

Reply:
I applied the `Service` approach here. I've there's a use case for overriding (besides cache wrapping) that can be dealt with in the future.